src/python.cpp

```diff
@@ -478,122 +478,24 @@ bool traceback_enabled() {
-
-  std::vector<std::string> traceback() const { return traceback_; };
-  void setTraceback(const std::vector<std::string>& traceback) {
-    traceback_ = traceback;
-  }
-
-  void clear() {
-    type_.clear();
-    value_.clear();
-    traceback_.clear();
-    message_.clear();
-  }
-
-private:
-  std::string type_;
-  std::string value_;
-  std::vector<std::string> traceback_;
-  std::string message_;
-};
-
-LastError s_lastError;
-
-//' Get or clear the last Python error encountered
-//'
-//' @return For `py_last_error()`, a list with the type, value,
-//' and traceback for the last Python error encountered (can be
-//' `NULL` if no error has yet been encountered).
-//'
-//' @export
-// [[Rcpp::export]]
-SEXP py_last_error() {
-  if (s_lastError.empty()) {
-    return R_NilValue;
-  } else {
-    List lastError;
-    lastError["type"] = s_lastError.type();
-    lastError["value"] = s_lastError.value();
-    lastError["traceback"] = s_lastError.traceback();
-    lastError["message"] = s_lastError.message();
-    return lastError;
-  }
-}
-
-//' @rdname py_last_error
-//' @export
-// [[Rcpp::export]]
-void py_clear_last_error() {
-  s_lastError.clear();
-}
-
-
-std::string py_fetch_error_type(const PyObjectPtr& pExcType) {
-
-  if (pExcType.is_null())
-    return std::string();
-
-  PyObjectPtr pStr(PyObject_GetAttrString(pExcType, "__name__"));
-  return as_std_string(pStr);
-
-}
-
-std::string py_fetch_error_value(const PyObjectPtr& pExcValue) {
-
-  if (pExcValue.is_null())
-    return std::string();
-
-  PyObjectPtr pStr(PyObject_Str(pExcValue));
-  return as_std_string(pStr);
-
-}
-
-void py_fetch_error_traceback(PyObject* excTraceback,
-                              std::vector<std::string>* pTraceback)
-{
-  if (excTraceback == NULL)
-    return;
-
-  // invoke 'traceback.format_tb(<traceback>)'
-  PyObjectPtr module(py_import("traceback"));
-  if (module.is_null())
-    return;
-
-  PyObjectPtr format_tb(PyObject_GetAttrString(module, "format_tb"));
-  if (format_tb.is_null())
-    return;
-
-  PyObjectPtr tb(PyObject_CallFunctionObjArgs(format_tb, excTraceback, NULL));
-  if (tb.is_null())
-    return;
-
-  // get the traceback
-  for (Py_ssize_t i = 0, n = PyList_Size(tb); i < n; i++)
-    pTraceback->push_back(as_std_string(PyList_GetItem(tb, i)));
-
-}
-
-// get a string representing the last python error
+// fetch and normalize the python exception object
+// save it in R reticulate:::.globals$py_last_exception
+// Return a short string suitable for Rcpp::stop(),
+// Either the exception message, or a truncated version with
+// user instructions for where to see the full exception.
 std::string py_fetch_error() {
+  PyObject *excType, *excValue, *excTraceback;
+  PyErr_Fetch(&excType, &excValue, &excTraceback);  // we now own the PyObjects
+  PyErr_NormalizeException(&excType, &excValue, &excTraceback);
+  if (excTraceback != NULL) {
+    PyException_SetTraceback(excValue, excTraceback);
```

Comment:
Can you confirm that this increases the Python reference count on excTraceback?

Reply:
[confirmed](https://github.com/python/cpython/blob/0575551f69ba9c999835bfb176a543d468083c03/Objects/exceptions.c#L273)